File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
  constructor(real, imag) {
    [this._real, this._imag] = [real, imag];
  }

  get real() {
    return this._real;
  }

  get imag() {
    return this._imag;
  }

  add(num2) {
    return new ComplexNumber(this.real + num2.real, this.imag + num2.imag);
  }

  sub(num2) {
    return new ComplexNumber(this.real - num2.real, this.imag - num2.imag);
  }

  div(num2) {
    let realPart = (this.real * num2.real + this.imag * num2.imag) / (num2.real ** 2 + num2.imag ** 2);
    let imagPart = (this.imag * num2.real - this.real * num2.imag) / (num2.real ** 2 + num2.imag ** 2);
    return new ComplexNumber(realPart, imagPart);
  }

  mul(num2) {
    return new ComplexNumber(this.real * num2.real - this.imag * num2.imag, this.real * num2.imag + this.imag * num2.real);
  }

  get abs() {
    return Math.sqrt(this.real ** 2 + this.imag ** 2);
  }

  get conj() {
    return new ComplexNumber(this.real, this.imag === 0 ? 0 : -this.imag);
  }

  get exp() {
    return new ComplexNumber(Math.E ** this.real * Math.cos(this.imag), Math.E ** this.real * Math.sin(this.imag));
  }
}

File: prolog/complex-numbers/complex_numbers.pl
real((R1, _), R) :- R = R1.
imaginary((_, I1), I) :- I = I1.

conjugate((R1, I1), Conjugate) :- Conjugate = (R1, -I1).
abs((R1, I1), Abs) :- Abs = sqrt(R1 * R1 + I1 * I1).

add((R1, I1), (R2, I2), CAdd) :- CAdd = (R1 + R2, I1 + I2).
sub((R1, I1), (R2, I2), CSub) :- CSub = (R1 - R2, I1 - I2).

mul((R1, I1), (R2, I2), CMul) :- CMul = (R1 * R2 - I1 * I2, R1 * I2 + R2 * I1).
div((R1, I1), (R2, I2), CDiv) :- CDiv = ((R1 * R2 + I1 * I2) / (R2 * R2 + I2 * I2), (R2 * I1 - R1 * I2) / (R2 * R2 + I2 * I2)).
